### PR TITLE
[go1.17] Build images for go1.17

### PR DESCRIPTION
**Pull request by justaugustus**
#### What type of PR is this?

/kind feature
/area dependency

#### What this PR does / why we need it:

- kube-cross: Drop non-legacy go1.15 variant
- kube-cross: Remove etcd from non-legacy builds (ref: https://github.com/kubernetes/release/pull/2124, https://github.com/kubernetes/release/pull/2123#discussion_r652788228)
- images: Enforce Debian codenames for Golang-based images
  When there are multiple image builds in flight both upstream and downstream,
  we can run into situations where a new Debian version becomes the default for
  image builds, which can have unintended side-effects across release branches.

  Here we use explicit pairings of Golang/Debian versions to retrieve images
  
  Example: `FROM golang:1.16.7-buster`
- kube-cross: Use OS codenames to construct clearer versions for images
  
  Uses the following nomenclature:
  `v<kubernetes-major>-go<go-major>-<os-codename>.<revision>`

  Example: `v1.23.0-go1.17-buster.0`

- [go1.17] Build images for go1.17
  - kube-cross:v1.23.0-go1.17-buster.0
  - go-runner:v2.3.1-go1.17-buster.0

ref:
- https://github.com/kubernetes/release/issues/2169
- https://kubernetes.slack.com/archives/C2C40FMNF/p1626226473096200

/assign @saschagrunert @puerco @cpanato
cc: @kubernetes/release-engineering @dims @dekkagaijin @dlorenc 

#### Special notes for your reviewer:

Image build presubmits may fail as [upstream s390x jobs](https://doi-janky.infosiftr.net/job/multiarch/job/s390x/job/golang/) appear as "unstable" at the time of this PR's creation. (cc: @mkumatag @tianon)

#### Does this PR introduce a user-facing change?

```release-note
- kube-cross: Drop non-legacy go1.15 variant
- kube-cross: Remove etcd from non-legacy builds
- images: Enforce Debian codenames for Golang-based images
  When there are multiple image builds in flight both upstream and downstream,
  we can run into situations where a new Debian version becomes the default for
  image builds, which can have unintended side-effects across release branches.

  Here we use explicit pairings of Golang/Debian versions to retrieve images

  Example: `FROM golang:1.16.7-buster`
- kube-cross: Use OS codenames to construct clearer versions for images
  
  Uses the following nomenclature:
  `v<kubernetes-major>-go<go-major>-<os-codename>.<revision>`

  Example: `v1.23.0-go1.17-buster.0`

- [go1.17] Build images for go1.17
  - kube-cross:v1.23.0-go1.17-buster.0
  - go-runner:v2.3.1-go1.17-buster.0
```

### CODE REVIEW

**Comment by dims** [comment]
/lgtm

🤞🏾

File: images/build/go-runner/variants.yaml [hunk 1]
```diff
@@ -1,11 +1,11 @@
 variants:
   go1.17-buster:
     CONFIG: 'go1.17-buster'
-    IMAGE_VERSION: 'v2.3.1-go1.17rc2-buster.0'
+    IMAGE_VERSION: 'v2.3.1-go1.17-buster.0'
```
Comment (justaugustus) [inline images/build/go-runner/variants.yaml:4]: We can look at doing at `bullseye` variant once it's available for distroless: https://github.com/GoogleContainerTools/distroless/issues/780

**Comment by justaugustus** [comment]
I _think_ we're waiting for this PR before the new upstream builds trigger: https://github.com/docker-library/official-images/pull/10739

**Comment by justaugustus** [comment]
> I _think_ we're waiting for this PR before the new upstream builds trigger: [docker-library/official-images#10739](https://github.com/docker-library/official-images/pull/10739)

Yep! From https://github.com/docker-library/official-images/pull/10739#issuecomment-899916032:
> **WARNING:** this will likely not be available for a bit since the build queue will be very full due to the `bullseye` release and thus a rebuild of all Debian based images (#10730)
> 
> Watch the jobs in https://doi-janky.infosiftr.net/job/multiarch/view/images/view/golang/ for the architecture of your choice to see when it gets build an pushed to the architecture-specific namespace (e.g. https://hub.docker.com/r/amd64/golang). Once it is there, it will be a short while later that it'll then be at https://hub.docker.com/_/golang in the multi-architecture manifest. See also our FAQ entry for [An image's source changed in Git, now what?](https://github.com/docker-library/faq/tree/73f10b0daf2fb8e7b38efaccc0e90b3510919d51#an-images-source-changed-in-git-now-what).

**Comment by justaugustus** [comment]
Potential flake:
/test pull-release-image-go-runner

**Comment by justaugustus** [comment]
Potential flake:
/test pull-release-image-kube-cross

**Comment by justaugustus** [comment]
One more time before I start debugging with purpose...
/test pull-release-image-kube-cross

**Comment by justaugustus** [comment]
Due to Debian bullseye and Golang 1.17 upstream image updates happening roughly in tandem, the latest image builds shifted to using bullseye, instead of buster by default (ref: https://github.com/kubernetes/release/pull/2218).

I'm combining this PR with https://github.com/kubernetes/release/pull/2210 and https://github.com/kubernetes/release/pull/2123 to tackle this all at once and minimize the image promotion churn.

/hold

**Comment by justaugustus** [comment]
(quota exhausted; don't attempt retests at the moment)

**Comment by justaugustus** [comment]
Possible [flake](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/release/2211/pull-release-image-go-runner/1430021827639382016):
/test pull-release-image-go-runner

**Comment by justaugustus** [comment]
Tests appear to be passing, but contexts are stuck due to the GCB build queue getting from previous runs:
/test-all

**Comment by justaugustus** [comment]
> I'm combining this PR with #2210 and #2123 to tackle this all at once and minimize the image promotion churn.

On second thought, I'm going to hold the proposed changes in #2210 for another PR since:
1. This PR has already gotten to be pretty large with the Debian codename enforcement
2. The GCB build queue is timing out sending this many jobs to build

Ready for review.
/hold cancel
cc: @kubernetes/release-engineering

**Review by cpanato — APPROVED** [review]
/lgtm

**Comment by justaugustus** [comment]
Timeout:
/test pull-release-image-go-runner

**Pull request merged**